### PR TITLE
docs: 466/hash-verification

### PR DESCRIPTION
**Pull request by hajjimo**

### CODE REVIEW

File: docs/src/content/docs/introduction/hash-verification.mdx [hunk 1]
```diff
@@ -0,0 +1,45 @@
+---
+title: Hash verification
+---
+
+import { CodeBlock, LinkOut } from '@interledger/docs-design-system'
+
+Once a resource owner (RO) authorizes a client software, the authorization server (AS) will redeirect the RO's [identify provider (IdP)](/introduction/idp/) to the finish URI. In order to secure this communication and verify that the redirect indeed emanated from the AS, the AS will provide a hash parameter in the request to the client's callback URI. The client **_must_** verify this hash.
```
Comment (melissahenderson) [inline docs/src/content/docs/introduction/hash-verification.mdx:7]: - Typo "redeirect"
- Change "identify provider" to identity provider

File: docs/src/content/docs/introduction/hash-verification.mdx [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+<CodeBlock title="SHA2 256-bit hash example">
+
+```http
+x-gguKWTj8rQf7d7i3w3UhzvuJ5bpOlKyAlVpLxBffY
+```
+
+</CodeBlock>
+
+For more information refer to the <LinkOut href='https://datatracker.ietf.org/doc/html/draft-ietf-gnap-core-protocol-20#name-calculating-the-interaction'> "Calculating the interaction hash"</LinkOut> section of the GNAP specification.
```
Comment (melissahenderson) [inline docs/src/content/docs/introduction/hash-verification.mdx:45]: Can you remove the quotation marks around the link name?

File: docs/src/content/docs/introduction/hash-verification.mdx [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+title: Hash verification
+---
+
+import { CodeBlock, LinkOut } from '@interledger/docs-design-system'
+
+Once a resource owner (RO) authorizes a client software, the authorization server (AS) will redirect the RO's [identity provider (IdP)](/introduction/idp/) to the finish URI. In order to secure this communication and verify that the redirect indeed emanated from the AS, the AS will provide a hash parameter in the request to the client's callback URI. The client **_must_** verify this hash.
+
+## Hashing method
+
+The hash base is generated by concatentating the following values in sequence using a single newline (0x0A) character to separate them:
```
Comment (raducristianpopa) [inline docs/src/content/docs/introduction/hash-verification.mdx:11]:
```suggestion
The hash base is generated by concatentating the following values in sequence using a single newline (\n) character to separate them:
```

File: docs/src/content/docs/introduction/hash-verification.mdx [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+
+1. `nonce` value sent by the client in the initial request.
+2. `nonce` value sent by the AS from the interaction finish response.
```
Comment (raducristianpopa) [inline docs/src/content/docs/introduction/hash-verification.mdx:14]: The interact nonce is returned after making the initial grant request (interaction instructions):

```json
{
  "interact": {
    "redirect": "https://auth.rafiki.money/4CF492MLVMSW9MKMXKHQ",
    "finish": "4105340a-05eb-4290-8739-f9e2b463bfa7" // <- nonce
  },
  "continue": {
    "access_token": {
      "value": "33OMUKMKSKU80UPRY5NM"
    },
    "uri": "https://auth.rafiki.money/continue/4CF492MLVMSW9MKMXKHQ",
    "wait": 30
  }
}
```

File: docs/src/content/docs/introduction/hash-verification.mdx [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+MBDOFXG4Y5CVJCX821LH
+4IFWWIKYB2PQ6U56NL1
+https://server.example.com/tx
+```
+
+</CodeBlock>
+
+The ASCII encoding of this string is hashed with the algorithm specified in the `hash_method` parameter under the `finish` key of the interaction finish request. The byte array from the hash function is then encoded using URL-Safe Base64 with no padding. The resultant string is the hash value.
+
+Unless specified by the client in the initial request, the `hash_method` will default to `sha-256`. If the client specifies the `hash_method`, the `hash_method` **_must_** be one of the hash name strings defined in the <LinkOut href='https://www.iana.org/assignments/named-information/named-information.xhtml#hash-alg'>IANA Named Information Hash Algorithm Registry</LinkOut>.
```
Comment (raducristianpopa) [inline docs/src/content/docs/introduction/hash-verification.mdx:33]: @mkurapov are we planning to support other hashing methods as well? If not I think we should only mention `sha-256`.

File: docs/src/content/docs/introduction/hash-verification.mdx [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+```http
+VJLO6A4CATR0KRO
+MBDOFXG4Y5CVJCX821LH
+4IFWWIKYB2PQ6U56NL1
+https://server.example.com/tx
+```
+
+</CodeBlock>
+
+The ASCII encoding of this string is hashed with the algorithm specified in the `hash_method` parameter under the `finish` key of the interaction finish request. The byte array from the hash function is then encoded using URL-Safe Base64 with no padding. The resultant string is the hash value.
```
Comment (raducristianpopa) [inline docs/src/content/docs/introduction/hash-verification.mdx:31]: In Rafiki we are doing only doing `base64` encoding not `base64url`.

CC @mkurapov

File: docs/src/content/docs/introduction/hash-verification.mdx (hunk 5 above)
Comment (mkurapov) [reply docs/src/content/docs/introduction/hash-verification.mdx:33]: I don't think so, we can just mention `sha-256`

File: docs/src/content/docs/introduction/hash-verification.mdx [hunk 7]
```diff
@@ -0,0 +1,45 @@
+---
+title: Hash verification
+---
+
+import { CodeBlock, LinkOut } from '@interledger/docs-design-system'
+
+Once a resource owner (RO) authorizes a client software, the authorization server (AS) will redirect the RO's [identity provider (IdP)](/introduction/idp/) to the finish URI. In order to secure this communication and verify that the redirect indeed emanated from the AS, the AS will provide a hash parameter in the request to the client's callback URI. The client **_must_** verify this hash.
```
[the remark this answers is not in the document]
Comment (mkurapov) [reply docs/src/content/docs/introduction/hash-verification.mdx:7]: I think we need to add a small section about how the client actually would do verification: by generating the hash using the components below, and comparing it to the provided value in the final redirect

File: docs/src/content/docs/introduction/idp.mdx [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+
+An identity provider (IdP) is a system or service that manages user authentication, identity information, and consent. Open Payments requires any authorization server (AS) that issues <Tooltip content="A grant requiring explicit interaction/consent from the resource owner before an access token can be issued" client:load><span>interactive grants</span></Tooltip> be integrated with an IdP.
+
+After an interactive grant request is initiated and the AS sets the session, the AS provides the client with the IdP URI to redirect the end-user to.
+
+<Interactive />
+
+:::tip[Reference implementation]
+Rafiki provides a reference <LinkOut href='https://github.com/interledger/rafiki/tree/main/packages/auth'>authorization service</LinkOut> implementation that includes support for integration with an IdP.
+:::
```
Comment (mkurapov) [inline docs/src/content/docs/introduction/idp.mdx:22]: Let's link to the Rafiki docs?

File: docs/src/content/docs/introduction/hash-verification.mdx [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+
+```http
+x-gguKWTj8rQf7d7i3w3UhzvuJ5bpOlKyAlVpLxBffY
+```
+
+</CodeBlock>
+
+## Verifying hash
+
+When the client receives a redirect from the AS, the AS will include the hash parameter in the response. The client must calculate this exact value by concatenating the fields referenced above and then applying the `sha-256` hashing algorithm. If the hash value matches the parameter sent by the AS, then the client can be certain the redirect emanated from the AS.
```
Comment (mkurapov) [inline docs/src/content/docs/introduction/hash-verification.mdx:45]: If you think its valuable, we can also add a JS example:

```js
function verifyHash(
  clientNonce,
  interactNonce,
  interactRef,
  authServerUrl,
  receivedHash
) {
  const data = `${clientNonce}\n${interactNonce}\n${interactRef}\n${authServerUrl}/`
  const hash = createHash('sha-256').update(data).digest('base64')

  return hash === receivedHash
}
```

**Review by mkurapov — APPROVED** [review]

**Pull request merged**